lib/services/local/generic/generic.go

```diff
@@ -0,0 +1,290 @@
+
+// Config is the configuration for the service configuration.
+type ServiceConfig[T types.Resource] struct {
+	Backend       backend.Backend
+	ResourceKind  string
+	Limit         int
+	BackendPrefix string
+	MarshalFunc   MarshalFunc[T]
+	UnmarshalFunc UnmarshalFunc[T]
+}
+
+func (c *ServiceConfig[T]) CheckAndSetDefaults() error {
+	if c.Backend == nil {
+		return trace.BadParameter("backend is missing")
+	}
+	if c.ResourceKind == "" {
+		return trace.BadParameter("resource kind is missing")
+	}
+	if c.Limit < 0 {
+		return trace.BadParameter("limit must be 0 or greater")
+	}
+	if c.BackendPrefix == "" {
+		return trace.BadParameter("backend prefix is missing")
+	}
+	if c.MarshalFunc == nil {
+		return trace.BadParameter("marshal func is missing")
+	}
+	if c.UnmarshalFunc == nil {
+		return trace.BadParameter("unmarshal func is missing")
+	}
+
+	return nil
+}
+
+// Service is a generic service for interacting with resources in the backend.
+type Service[T types.Resource] struct {
+	backend       backend.Backend
+	resourceKind  string
+	limit         int
+	backendPrefix string
+	marshalFunc   MarshalFunc[T]
+	unmarshalFunc UnmarshalFunc[T]
+}
+
+// NewService will return a new generic service with the given config. This will
+// panic if the configuration is invalid.
+func NewService[T types.Resource](cfg *ServiceConfig[T]) *Service[T] {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		panic(fmt.Sprintf("Developer misconfiguration of generic service: %v", err))
+	}
+
+	return &Service[T]{
+		backend:       cfg.Backend,
+		resourceKind:  cfg.ResourceKind,
+		limit:         cfg.Limit,
+		backendPrefix: cfg.BackendPrefix,
+		marshalFunc:   cfg.MarshalFunc,
+		unmarshalFunc: cfg.UnmarshalFunc,
+	}
+}
+
+// GetResources returns a list of all resources.
+func (s *Service[T]) GetResources(ctx context.Context) ([]T, error) {
+	rangeStart := backend.Key(s.backendPrefix)
+	rangeEnd := backend.RangeEnd(backend.Key(s.backendPrefix, ""))
+
+	// no filter provided get the range directly
+	result, err := s.backend.GetRange(ctx, rangeStart, rangeEnd, backend.NoLimit)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	out := make([]T, 0, len(result.Items))
+	for _, item := range result.Items {
+		resource, err := s.unmarshalFunc(item.Value)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		out = append(out, resource)
+	}
+
+	return out, nil
+}
+
+// ListResources returns a paginated list of resources.
+func (s *Service[T]) ListResources(ctx context.Context, pageSize int, pageToken string) ([]T, string, error) {
+	rangeStart := backend.Key(s.backendPrefix, pageToken)
+	rangeEnd := backend.RangeEnd(backend.Key(s.backendPrefix, ""))
+
+	// Adjust page size, so it can't be too large.
+	if pageSize <= 0 || pageSize > s.limit {
+		pageSize = s.limit
+	}
+
+	// If backend.NoLimit is requested and the service permits it, set the limit to NoLimit.
+	limit := pageSize + 1
```

Comment:
What's supposed to happen here if the service's `Limit` is 0?

Reply:
Argh, I meant to get rid of that! Also, it looks like it's dependent on the backend implementation, but it seems to use a default limit:

```
	if limit <= 0 {
		limit = backend.DefaultRangeLimit
	}
```

It looks like each implementation does this. This seems reasonable, IMO.

Reply:
We should not be adding 1 to the page size if the page size is "infinite" (0) then.

Reply:
Okay, I've made a decision to make the page limit `backend.DefaultRangeLimit` if the limit is set to 0. This makes the pagination logic simpler.